### PR TITLE
CI: Upgrade ``array-api-tests``

### LINKED ISSUE #27072: BUG: astype() function does not work on scalars
### Describe the issue:

The new `astype()` function does not work on NumPy scalars. 

```
>>> np.astype(np.float32(0.), np.float64)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/numpy/numpy/_core/numeric.py", line 2688, in astype
    raise TypeError(
TypeError: Input should be a NumPy array. It is a <class 'numpy.float32'> instead.
>>> np.float32(0.).astype(np.float64)
np.float64(0.0)
```

Note that this breaks with array API compatibility. 

CC @mtsokol 

### Reproduce the code example:

```python
>>> np.astype(np.float32(0.), np.float64)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/numpy/numpy/_core/numeric.py", line 2688, in astype
    raise TypeError(
TypeError: Input should be a NumPy array. It is a <class 'numpy.float32'> instead.
>>> np.float32(0.).astype(np.float64)
np.float64(0.0)
```

### Error message:

_No response_

### Python and NumPy Versions:

2.1.0.dev0+git20240728.fdcbd0e
3.12.4 | packaged by conda-forge | (main, Jun 17 2024, 10:13:44) [Clang 16.0.6 ]

### Runtime Environment:

_No response_

### Context for the issue:

This leads to test failures in the array-api-tests test suite (test_unique_inverse and test_unique_values). If you are running those tests on CI you may want to investigate whey you didn't notice this earlier.

### PR DESCRIPTION
**Pull request by mtsokol**
Fixes https://github.com/numpy/numpy/issues/27072 and increases `array-api-tests` suite `--max-examples` to 50.

### CODE REVIEW

File: tools/ci/array-api-xfails.txt [hunk 1]
```diff
@@ -7,10 +7,14 @@ array_api_tests/test_signatures.py::test_func_signature[reshape]
...
 array_api_tests/test_signatures.py::test_func_signature[clip]
 
-# missing 'descending' keyword arguments
+# missing 'descending' keyword argument
 array_api_tests/test_signatures.py::test_func_signature[argsort]
 array_api_tests/test_signatures.py::test_func_signature[sort]
 
+# missing 'descending' keyword argument
+array_api_tests/test_sorting_functions.py::test_argsort
+array_api_tests/test_sorting_functions.py::test_sort
```
Comment (asmeurer) [inline tools/ci/array-api-xfails.txt:16]: The rest of these are bugs in the test suite but this looks like one thing that is actually missing from NumPy for array API compatibility.

**Comment by ngoldbaum** [comment]
The tests only take 26 seconds to run - maybe a value higher than 50? I don't have a good intuition about what a reasonable number to use here is, or whether it should be time-bound instead.

File: tools/ci/array-api-xfails.txt (hunk 1 above)
Comment (mtsokol) [reply tools/ci/array-api-xfails.txt:16]: I agree, but we decided that it would be too much churn to include it now.

**Comment by asmeurer** [comment]
The default value, which comes from hypothesis, is 100.

**Comment by mtsokol** [comment]
I increased `max-examples` to 100 and there's one more test failure:
```
FAILED array_api_tests/test_operators_and_elementwise_functions.py::test_clip - AssertionError: out.dtype=uint32, but should be uint16 [clip(uint16)]
Falsifying example: test_clip(
    x=array(0, dtype=uint16),
    data=data(...),
)
Draw 1 (min.shape, max.shape): ((), ())
Draw 2 (min): 0
Draw 3 (max): array(0, dtype=uint32)
Draw 4 (kwargs): {'min': 0, 'max': array(0, dtype=uint32)}
```
And I wonder if it's more an issue on the `array-api-tests` side. The standard says:
> If `x` and either `min` or `max` have different data type kinds (e.g., integer versus floating-point), behavior is unspecified and thus implementation-dependent.

And the failing example is:
```python
In [30]: x = np.array(0, dtype=np.uint16)

In [31]: np.clip(x, min=0, max=np.uint32(0))
Out[31]: np.uint32(0)
```
I think promotion here is intentional and if mixed data types of `x` and `min`/`max` is implementation-dependent then `array-api-tests` shouldn't require `x.dtype == np.clip(x, ...).dtype` here. WDYT?

**Comment by ngoldbaum** [comment]
On a hunch I just tried `max_examples=1000`, which also returned:

```
FAILED array_api_tests/test_statistical_functions.py::test_sum - AssertionError: out=2199023255552.0, but should be roughly 3298534883328.0 [sum()]
```

<details>

```
    @pytest.mark.unvectorized
>   @given(
        x=hh.arrays(
            dtype=hh.numeric_dtypes,
            shape=hh.shapes(min_side=1),
            elements={"allow_nan": False},
        ),
        data=st.data(),
    )

array_api_tests/test_statistical_functions.py:269:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = array([ 1.8446743e+19,  0.0000000e+00,  1.8446742e+19,  0.0000000e+00,
       -1.8446741e+19, -1.8446741e+19], dtype=float32), data = data(...)

    @pytest.mark.unvectorized
    @given(
        x=hh.arrays(
            dtype=hh.numeric_dtypes,
            shape=hh.shapes(min_side=1),
            elements={"allow_nan": False},
        ),
        data=st.data(),
    )
    def test_sum(x, data):
        kw = data.draw(
            hh.kwargs(
                axis=hh.axes(x.ndim),
                dtype=kwarg_dtypes(x.dtype),
                keepdims=st.booleans(),
            ),
            label="kw",
        )
        keepdims = kw.get("keepdims", False)

        with hh.reject_overflow():
            out = xp.sum(x, **kw)

        dtype = kw.get("dtype", None)
        expected_dtype = dh.accumulation_result_dtype(x.dtype, dtype)
        if expected_dtype is None:
            # If a default uint cannot exist (i.e. in PyTorch which doesn't support
            # uint32 or uint64), we skip testing the output dtype.
            # See https://github.com/data-apis/array-api-tests/issues/160
            if x.dtype in dh.uint_dtypes:
                assert dh.is_int_dtype(out.dtype)  # sanity check
        else:
            ph.assert_dtype("sum", in_dtype=x.dtype, out_dtype=out.dtype, expected=expected_dtype)
        _axes = sh.normalize_axis(kw.get("axis", None), x.ndim)
        ph.assert_keepdimable_shape(
            "sum", in_shape=x.shape, out_shape=out.shape, axes=_axes, keepdims=keepdims, kw=kw
        )
        scalar_type = dh.get_scalar_type(out.dtype)
        for indices, out_idx in zip(sh.axes_ndindex(x.shape, _axes), sh.ndindex(out.shape)):
            sum_ = scalar_type(out[out_idx])
            assume(cmath.isfinite(sum_))
            elements = []
            for idx in indices:
                s = scalar_type(x[idx])
                elements.append(s)
            expected = sum(elements)
            if dh.is_int_dtype(out.dtype):
                m, M = dh.dtype_ranges[out.dtype]
                assume(m <= expected <= M)
>           ph.assert_scalar_equals("sum", type_=scalar_type, idx=out_idx, out=sum_, expected=expected)
E           AssertionError: out=2199023255552.0, but should be roughly 3298534883328.0 [sum()]
E           Falsifying example: test_sum(
E               x=array([ 1.8446743e+19,  0.0000000e+00,  1.8446742e+19,  0.0000000e+00,
E                      -1.8446741e+19, -1.8446741e+19], dtype=float32),
E               data=data(...),
E           )
E           Draw 1 (kw): {'axis': None}

array_api_tests/test_statistical_functions.py:317: AssertionError
```
</details>

**Comment by asmeurer** [comment]
The sum thing is indeed a known issue with the test suite. Here's a more accurate set of xfails for numpy https://github.com/data-apis/array-api-compat/blob/main/numpy-dev-xfails.txt.

The `clip` thing is correct. The standard says that `clip` should return the same dtype as the first argument. The text you quoted is irrelevant because there is no mixing of data type kinds in that example (everything is integer). See https://github.com/numpy/numpy/issues/24976 for more details.

**Comment by ngoldbaum** [comment]
Maybe you should be using the skips file maintained by the array-api-tests repo like Aaron suggested? Since it captures issues discovered upstream.

**Comment by mtsokol** [comment]
> Maybe you should be using the skips file maintained by the array-api-tests repo like Aaron suggested? Since it captures issues discovered upstream.

In this PR I'm using `--xfails-file` feature instead of `--skips-file`. This means `tools/ci/array-api-xfails.txt` tests are run and verified that they fail. If a particular test passes then the test suite reports an `XPASS` error. Therefore I can't extent the xfails files with `array-api-tests` counterpart as we would xfail passing tests.

I think it's more accurate to use `--xfails-file` because we make sure that these tests fail indeed instead of just skipping them.

**Comment by ngoldbaum** [comment]
Sorry, I mean the xfails
file maintained upstream: https://github.com/data-apis/array-api-compat/blob/main/numpy-dev-xfails.txt

**Comment by mtsokol** [comment]
I think this file is outdated, it contains non-existing tests. Here I updated it: https://github.com/data-apis/array-api-compat/pull/171

**Comment by ngoldbaum** [comment]
If you can't just immediately use that let's move to using the upstream file in a followup. Thanks @mtsokol!

**Pull request merged**